File: tools/nia_tools/audit/compatibility.py
#!/usr/bin/env python3
"""Audit Nia compatibility identities and fingerprint domain spelling."""

from __future__ import annotations

import argparse
import ast
import os
import re
import tomllib
from collections.abc import Sequence
from pathlib import Path

from tools.nia_tools.repository import REPOSITORY_ROOT

ROOT = REPOSITORY_ROOT
REGISTRY = Path("crates/nia-compat/src/lib.rs")
DOMAIN_IDENTITY = (
    r"nia\.[a-z0-9]+(?:-[a-z0-9]+)*"
    r"(?:\.[a-z0-9]+(?:-[a-z0-9]+)*)*\.v[1-9][0-9]*"
)
DOMAIN = re.compile(rf"^{DOMAIN_IDENTITY}$")
STRING = re.compile(r'"((?:\\.|[^"\\\r\n])*)"')
CONSTRUCTOR_DOMAIN = re.compile(
    r"(?:QueryFingerprintBuilder|Encoder)::new\s*\(\s*"
    r'"((?:\\.|[^"\\\r\n])*)"',
    re.DOTALL,
)
TYPED_DOMAIN_DECLARATION = re.compile(
    r"\bconst\s+([A-Z][A-Z0-9_]*)\s*:\s*FingerprintDomain\s*=\s*"
    r"FingerprintDomain::new\s*\(\s*"
    rf'"({DOMAIN_IDENTITY})"\s*\)',
    re.DOTALL,
)
BYTE_STRING = re.compile(r'b"((?:\\.|[^"\\\r\n])*)"')
FORBIDDEN_IDENTITY_NAMES = (
    "RESOURCE_LAYOUT_SCHEMA",
    "STD_SCHEMA",
    "BUILD_PROTOCOL_SCHEMA",
    "BUILD_PLAN_SCHEMA_VERSION",
    "MANGLE_ABI_VERSION",
    "CODEGEN_ABI_VERSION",
    "RUNNER_CONFIG_SCHEMA_VERSION",
    "RUNNER_CONFIG_MAGIC",
    "RUNNER_CONFIG_MAGIC_TEXT",
    "FRONTEND_CACHE_SCHEMA",
    "OBJECT_WORK_PRODUCT_SCHEMA",
    "LINK_RESULT_SCHEMA",
    "ARCHIVE_SCHEMA",
)
TEXT_SUFFIXES = {
    ".json",
    ".md",
    ".meta",
    ".nia",
    ".py",
    ".rs",
    ".sh",
    ".toml",
    ".txt",
    ".yaml",
    ".yml",
}
VERSION_AUTHORITIES = {
    Path("Cargo.toml"),
    Path("Cargo.lock"),
    Path("lib/toolchain.meta"),
}


def source_files(root: Path) -> list[Path]:
    files = []
    for directory, names, filenames in os.walk(root):
        names[:] = [name for name in names if name not in {".git", "target"}]
        files.extend(Path(directory) / filename for filename in filenames)
    return files


def rust_sources(root: Path) -> list[Path]:
    crates = root / "crates"
    return sorted(crates.glob("*/src/**/*.rs")) if crates.is_dir() else []


def production_rust_sources(root: Path) -> list[Path]:
    return [
        path
        for path in rust_sources(root)
        if path.name != "tests.rs" and "tests" not in path.relative_to(root).parts
    ]


def production_source(path: Path) -> str:
    source = path.read_text(encoding="utf-8")
    test_module = re.search(r"#\[cfg\(test\)\]\s*mod\s+tests\s*\{", source)
    return source[: test_module.start()] if test_module else source


def decode_string(value: str) -> str:
    return value


def registered_magics(root: Path) -> set[bytes]:
    source = (root / REGISTRY).read_text(encoding="utf-8")
    magics = set()
    for match in BYTE_STRING.finditer(source):
        if not match.group(1).startswith("NIA"):
            continue
        value = ast.literal_eval(f'b"{match.group(1)}"')
        if len(value) == 8:
            magics.add(value)
    return magics


def fingerprint_domain_errors(root: Path) -> list[str]:
    errors = []
    for path in production_rust_sources(root):
        source = production_source(path)
        relative = path.relative_to(root)
        checked = set()
        for match in CONSTRUCTOR_DOMAIN.finditer(source):
            domain = decode_string(match.group(1))
            checked.add((match.start(1), domain))
            if not DOMAIN.fullmatch(domain):
                errors.append(f"{relative}: invalid fingerprint domain `{domain}`")
        for match in STRING.finditer(source):
            value = decode_string(match.group(1))
            if not value.startswith("nia.") or value == "nia.compiler_builtins":
                continue
            if not DOMAIN.fullmatch(value) and (match.start(1), value) not in checked:
                errors.append(f"{relative}: invalid Nia identity string `{value}`")
    return errors


def typed_fingerprint_domain_errors(root: Path) -> list[str]:
    errors = []
    declarations: dict[str, list[str]] = {}
    for path in production_rust_sources(root):
        source = production_source(path)
        relative = path.relative_to(root)
        declaration_spans = []
        for match in TYPED_DOMAIN_DECLARATION.finditer(source):
            declaration_spans.append(match.span(2))
            declarations.setdefault(match.group(2), []).append(
                f"{relative}:{source.count(chr(10), 0, match.start()) + 1}"
            )
        for match in CONSTRUCTOR_DOMAIN.finditer(source):
            errors.append(
                f"{relative}: fingerprint builder receives raw literal `{match.group(1)}`"
            )
        for match in STRING.finditer(source):
            value = match.group(1)
            if not DOMAIN.fullmatch(value):
                continue
            if not any(start <= match.start(1) < end for start, end in declaration_spans):
                errors.append(
                    f"{relative}: fingerprint domain `{value}` is not an owner-local typed constant"
                )
    for domain, owners in declarations.items():
        if len(owners) > 1:
            errors.append(
                f"duplicate fingerprint domain `{domain}` declared at {', '.join(owners)}"
            )
    return errors


def global_identity_errors(root: Path) -> list[str]:
    errors = []
    registry_path = root / REGISTRY
    magics = registered_magics(root)
    for path in rust_sources(root):
        if path == registry_path:
            continue
        source = path.read_text(encoding="utf-8")
        relative = path.relative_to(root)
        for match in BYTE_STRING.finditer(source):
            if not match.group(1).startswith("NIA"):
                continue
            value = ast.literal_eval(f'b"{match.group(1)}"')
            if value in magics:
                errors.append(f"{relative}: registered magic is defined outside nia-compat")
        for name in FORBIDDEN_IDENTITY_NAMES:
            if re.search(rf"\b(?:pub(?:\([^)]*\))?\s+)?const\s+{name}\b", source):
                errors.append(f"{relative}: obsolete compatibility identity `{name}`")
        if "CARGO_PKG_VERSION" in source:
            errors.append(f"{relative}: release version is read outside nia-compat")
    return errors


def workspace_version(root: Path) -> str:
    manifest = tomllib.loads((root / "Cargo.toml").read_text(encoding="utf-8"))
    return manifest["workspace"]["package"]["version"]


def release_version_errors(root: Path) -> list[str]:
    version = workspace_version(root)
    errors = []
    for path in source_files(root):
        relative = path.relative_to(root)
        if relative in VERSION_AUTHORITIES or path.suffix not in TEXT_SUFFIXES:
            continue
        try:
            source = path.read_text(encoding="utf-8")
        except UnicodeDecodeError:
            continue
        if version in source:
            errors.append(f"{relative}: workspace version `{version}` is duplicated")
    return errors


def audit(root: Path = ROOT) -> list[str]:
    return [
        *fingerprint_domain_errors(root),
        *typed_fingerprint_domain_errors(root),
        *global_identity_errors(root),
        *release_version_errors(root),
    ]


def parse_args(arguments: Sequence[str] | None = None) -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        prog="python3 -m tools audit compatibility", description=__doc__
    )
    parser.add_argument("--root", type=Path, default=ROOT)
    return parser.parse_args(arguments)


def main(arguments: Sequence[str] | None = None) -> int:
    errors = audit(parse_args(arguments).root.resolve())
    if errors:
        raise SystemExit("compatibility audit failed:\n" + "\n".join(errors))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/nia-maintain/src/audit/compatibility.rs
use std::collections::{BTreeMap, BTreeSet};
use std::fs;
use std::path::{Path, PathBuf};
use std::sync::LazyLock;

use regex::Regex;

use crate::MaintainResult;

const REGISTRY: &str = "crates/nia-compat/src/lib.rs";
const DOMAIN_IDENTITY: &str =
    r"nia\.[a-z0-9]+(?:-[a-z0-9]+)*(?:\.[a-z0-9]+(?:-[a-z0-9]+)*)*\.v[1-9][0-9]*";
const FORBIDDEN_IDENTITY_NAMES: [&str; 13] = [
    "RESOURCE_LAYOUT_SCHEMA",
    "STD_SCHEMA",
    "BUILD_PROTOCOL_SCHEMA",
    "BUILD_PLAN_SCHEMA_VERSION",
    "MANGLE_ABI_VERSION",
    "CODEGEN_ABI_VERSION",
    "RUNNER_CONFIG_SCHEMA_VERSION",
    "RUNNER_CONFIG_MAGIC",
    "RUNNER_CONFIG_MAGIC_TEXT",
    "FRONTEND_CACHE_SCHEMA",
    "OBJECT_WORK_PRODUCT_SCHEMA",
    "LINK_RESULT_SCHEMA",
    "ARCHIVE_SCHEMA",
];
const TEXT_SUFFIXES: [&str; 12] = [
    "json", "md", "meta", "nia", "py", "rs", "sh", "toml", "txt", "yaml", "yml", "lock",
];
const VERSION_AUTHORITIES: [&str; 3] = ["Cargo.toml", "Cargo.lock", "lib/toolchain.meta"];

static DOMAIN: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(&format!(r"^{DOMAIN_IDENTITY}$")).expect("valid domain regex"));
static STRING: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r#""((?:\\.|[^"\\\r\n])*)""#).expect("valid string regex"));
static CONSTRUCTOR_DOMAIN: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r#"(?s)(?:QueryFingerprintBuilder|Encoder)::new\s*\(\s*"((?:\\.|[^"\\\r\n])*)""#)
        .expect("valid constructor domain regex")
});
static TYPED_DOMAIN_DECLARATION: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(&format!(
        r#"(?s)\bconst\s+([A-Z][A-Z0-9_]*)\s*:\s*FingerprintDomain\s*=\s*FingerprintDomain::new\s*\(\s*"({DOMAIN_IDENTITY})"\s*\)"#
    ))
    .expect("valid typed domain regex")
});
static BYTE_STRING: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r#"b"((?:\\.|[^"\\\r\n])*)""#).expect("valid byte string regex"));
static TEST_MODULE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"#\[cfg\(test\)\]\s*mod\s+tests\s*\{").expect("valid test module regex")
});
static FORBIDDEN_IDENTITY: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(&format!(
        r"\b(?:pub(?:\([^)]*\))?\s+)?const\s+({})\b",
        FORBIDDEN_IDENTITY_NAMES.join("|")
    ))
    .expect("valid forbidden identity regex")
});

fn collect_files(root: &Path, files: &mut Vec<PathBuf>) -> MaintainResult<()> {
    let entries = fs::read_dir(root)
        .map_err(|error| format!("failed to read directory {}: {error}", root.display()))?;
    for entry in entries {
        let entry = entry.map_err(|error| {
            format!(
                "failed to inspect directory entry in {}: {error}",
                root.display()
            )
        })?;
        let file_type = entry
            .file_type()
            .map_err(|error| format!("failed to inspect {}: {error}", entry.path().display()))?;
        let path = entry.path();
        if file_type.is_dir() {
            let name = entry.file_name();
            if matches!(
                name.to_str(),
                Some(".git" | "target" | "node_modules" | "__pycache__")
            ) {
                continue;
            }
            collect_files(&path, files)?;
        } else if file_type.is_file() {
            files.push(path);
        }
    }
    Ok(())
}

fn source_files(root: &Path) -> MaintainResult<Vec<PathBuf>> {
    let mut files = Vec::new();
    collect_files(root, &mut files)?;
    files.sort();
    Ok(files)
}

fn rust_sources(root: &Path) -> MaintainResult<Vec<PathBuf>> {
    let crates = root.join("crates");
    if !crates.is_dir() {
        return Ok(Vec::new());
    }
    let mut files = Vec::new();
    for entry in fs::read_dir(&crates)
        .map_err(|error| format!("failed to read {}: {error}", crates.display()))?
    {
        let crate_path = entry
            .map_err(|error| format!("failed to inspect {}: {error}", crates.display()))?
            .path();
        let source = crate_path.join("src");
        if source.is_dir() {
            collect_files(&source, &mut files)?;
        }
    }
    files.retain(|path| path.extension().is_some_and(|extension| extension == "rs"));
    files.sort();
    Ok(files)
}

fn production_rust_sources(root: &Path) -> MaintainResult<Vec<PathBuf>> {
    Ok(rust_sources(root)?
        .into_iter()
        .filter(|path| {
            path.file_name().is_none_or(|name| name != "tests.rs")
                && !path.strip_prefix(root).is_ok_and(|relative| {
                    relative
                        .components()
                        .any(|part| part.as_os_str() == "tests")
                })
        })
        .collect())
}

fn production_source(path: &Path) -> MaintainResult<String> {
    let source = fs::read_to_string(path)
        .map_err(|error| format!("failed to read {}: {error}", path.display()))?;
    Ok(TEST_MODULE.find(&source).map_or(source.clone(), |matched| {
        source[..matched.start()].to_owned()
    }))
}

fn relative(root: &Path, path: &Path) -> String {
    path.strip_prefix(root)
        .unwrap_or(path)
        .to_string_lossy()
        .replace('\\', "/")
}

fn decode_byte_string(value: &str) -> Option<Vec<u8>> {
    let bytes = value.as_bytes();
    let mut decoded = Vec::with_capacity(bytes.len());
    let mut index = 0;
    while index < bytes.len() {
        if bytes[index] != b'\\' {
            decoded.push(bytes[index]);
            index += 1;
            continue;
        }
        index += 1;
        let escaped = *bytes.get(index)?;
        match escaped {
            b'\\' | b'"' | b'\'' => decoded.push(escaped),
            b'n' => decoded.push(b'\n'),
            b'r' => decoded.push(b'\r'),
            b't' => decoded.push(b'\t'),
            b'0' => decoded.push(0),
            b'x' => {
                let high = *bytes.get(index + 1)?;
                let low = *bytes.get(index + 2)?;
                let digit = |byte: u8| match byte {
                    b'0'..=b'9' => Some(byte - b'0'),
                    b'a'..=b'f' => Some(byte - b'a' + 10),
                    b'A'..=b'F' => Some(byte - b'A' + 10),
                    _ => None,
                };
                decoded.push(digit(high)? * 16 + digit(low)?);
                index += 2;
            }
            _ => return None,
        }
        index += 1;
    }
    Some(decoded)
}

fn registered_magics(root: &Path) -> MaintainResult<BTreeSet<Vec<u8>>> {
    let registry = root.join(REGISTRY);
    let source = fs::read_to_string(&registry)
        .map_err(|error| format!("failed to read {}: {error}", registry.display()))?;
    Ok(BYTE_STRING
        .captures_iter(&source)
        .filter_map(|captures| decode_byte_string(&captures[1]))
        .filter(|value| value.starts_with(b"NIA") && value.len() == 8)
        .collect())
}

pub fn fingerprint_domain_errors(root: &Path) -> MaintainResult<Vec<String>> {
    let mut errors = Vec::new();
    for path in production_rust_sources(root)? {
        let source = production_source(&path)?;
        let path = relative(root, &path);
        let mut checked = BTreeSet::new();
        for captures in CONSTRUCTOR_DOMAIN.captures_iter(&source) {
            let matched = captures.get(1).expect("constructor domain capture");
            let domain = matched.as_str();
            checked.insert((matched.start(), domain.to_owned()));
            if !DOMAIN.is_match(domain) {
                errors.push(format!("{path}: invalid fingerprint domain `{domain}`"));
            }
        }
        for captures in STRING.captures_iter(&source) {
            let matched = captures.get(1).expect("string capture");
            let value = matched.as_str();
            if !value.starts_with("nia.") || value == "nia.compiler_builtins" {
                continue;
            }
            if !DOMAIN.is_match(value) && !checked.contains(&(matched.start(), value.to_owned())) {
                errors.push(format!("{path}: invalid Nia identity string `{value}`"));
            }
        }
    }
    Ok(errors)
}

pub fn typed_fingerprint_domain_errors(root: &Path) -> MaintainResult<Vec<String>> {
    let mut errors = Vec::new();
    let mut declarations = BTreeMap::<String, Vec<String>>::new();
    for path in production_rust_sources(root)? {
        let source = production_source(&path)?;
        let path = relative(root, &path);
        let mut declaration_spans = Vec::new();
        for captures in TYPED_DOMAIN_DECLARATION.captures_iter(&source) {
            let domain = captures.get(2).expect("typed domain capture");
            declaration_spans.push(domain.start()..domain.end());
            let line = source[..captures.get(0).expect("declaration capture").start()]
                .bytes()
                .filter(|byte| *byte == b'\n')
                .count()
                + 1;
            declarations
                .entry(domain.as_str().to_owned())
                .or_default()
                .push(format!("{path}:{line}"));
        }
        for captures in CONSTRUCTOR_DOMAIN.captures_iter(&source) {
            errors.push(format!(
                "{path}: fingerprint builder receives raw literal `{}`",
                &captures[1]
            ));
        }
        for captures in STRING.captures_iter(&source) {
            let matched = captures.get(1).expect("string capture");
            let value = matched.as_str();
            if DOMAIN.is_match(value)
                && !declaration_spans
                    .iter()
                    .any(|span| span.contains(&matched.start()))
            {
                errors.push(format!(
                    "{path}: fingerprint domain `{value}` is not an owner-local typed constant"
                ));
            }
        }
    }
    for (domain, owners) in declarations {
        if owners.len() > 1 {
            errors.push(format!(
                "duplicate fingerprint domain `{domain}` declared at {}",
                owners.join(", ")
            ));
        }
    }
    Ok(errors)
}

pub fn global_identity_errors(root: &Path) -> MaintainResult<Vec<String>> {
    let mut errors = Vec::new();
    let registry = root.join(REGISTRY);
    let magics = registered_magics(root)?;
    for path in rust_sources(root)? {
        if path == registry {
            continue;
        }
        let source = fs::read_to_string(&path)
            .map_err(|error| format!("failed to read {}: {error}", path.display()))?;
        let relative = relative(root, &path);
        for captures in BYTE_STRING.captures_iter(&source) {
            if let Some(value) = decode_byte_string(&captures[1])
                && value.starts_with(b"NIA")
                && magics.contains(&value)
            {
                errors.push(format!(
                    "{relative}: registered magic is defined outside nia-compat"
                ));
            }
        }
        for captures in FORBIDDEN_IDENTITY.captures_iter(&source) {
            errors.push(format!(
                "{relative}: obsolete compatibility identity `{}`",
                &captures[1]
            ));
        }
        if source.contains("CARGO_PKG_VERSION") {
            errors.push(format!(
                "{relative}: release version is read outside nia-compat"
            ));
        }
    }
    Ok(errors)
}

fn workspace_version(root: &Path) -> MaintainResult<String> {
    let path = root.join("Cargo.toml");
    let source = fs::read_to_string(&path)
        .map_err(|error| format!("failed to read {}: {error}", path.display()))?;
    let manifest: toml::Value = toml::from_str(&source)
        .map_err(|error| format!("failed to parse {}: {error}", path.display()))?;
    manifest
        .get("workspace")
        .and_then(|workspace| workspace.get("package"))
        .and_then(|package| package.get("version"))
        .and_then(toml::Value::as_str)
        .map(str::to_owned)
        .ok_or_else(|| "Cargo.toml has no workspace.package.version".to_owned())
}

pub fn release_version_errors(root: &Path) -> MaintainResult<Vec<String>> {
    let version = workspace_version(root)?;
    let authorities = VERSION_AUTHORITIES.into_iter().collect::<BTreeSet<_>>();
    let suffixes = TEXT_SUFFIXES.into_iter().collect::<BTreeSet<_>>();
    let mut errors = Vec::new();
    for path in source_files(root)? {
        let relative = relative(root, &path);
        let suffix = path.extension().and_then(|value| value.to_str());
        if authorities.contains(relative.as_str()) || !suffix.is_some_and(|s| suffixes.contains(s))
        {
            continue;
        }
        let Ok(source) = fs::read_to_string(&path) else {
            continue;
        };
        if source.contains(&version) {
            errors.push(format!(
                "{relative}: workspace version `{version}` is duplicated"
            ));
        }
    }
    Ok(errors)
}

pub fn audit(root: &Path) -> MaintainResult<Vec<String>> {
    let mut errors = fingerprint_domain_errors(root)?;
    errors.extend(typed_fingerprint_domain_errors(root)?);
    errors.extend(global_identity_errors(root)?);
    errors.extend(release_version_errors(root)?);
    Ok(errors)
}

pub fn run(root: &Path) -> MaintainResult<()> {
    let errors = audit(root)?;
    if errors.is_empty() {
        Ok(())
    } else {
        Err(format!(
            "compatibility audit failed:\n{}",
            errors.join("\n")
        ))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_support::TestDirectory;

    fn repository(name: &str) -> TestDirectory {
        let directory = TestDirectory::new(name);
        directory.write(
            "Cargo.toml",
            "[workspace]\nmembers = []\n[workspace.package]\nversion = \"1.2.3\"\n",
        );
        directory.write(
            "crates/nia-compat/src/lib.rs",
            "pub const FORMAT: &[u8; 8] = b\"NIAFMT01\";\n",
        );
        directory.write("crates/owner/src/lib.rs", "");
        directory
    }

    #[test]
    fn accepts_versioned_fingerprint_domains() {
        let directory = repository("valid-domain");
        directory.write(
            "crates/owner/src/lib.rs",
            "const PRODUCT_DOMAIN: FingerprintDomain =\n    FingerprintDomain::new(\"nia.owner.product.v2\");\nQueryFingerprintBuilder::new(PRODUCT_DOMAIN);\n",
        );
        assert!(
            fingerprint_domain_errors(directory.path())
                .unwrap()
                .is_empty()
        );
        assert!(
            typed_fingerprint_domain_errors(directory.path())
                .unwrap()
                .is_empty()
        );
    }

    #[test]
    fn rejects_unversioned_and_raw_constructor_domains() {
        let directory = repository("raw-domain");
        directory.write(
            "crates/owner/src/lib.rs",
            "QueryFingerprintBuilder::new(\"nia.owner.product.v2\");\nQueryFingerprintBuilder::new(\"owner-product\");\n",
        );
        assert!(fingerprint_domain_errors(directory.path()).unwrap()[0].contains("owner-product"));
        assert!(
            typed_fingerprint_domain_errors(directory.path()).unwrap()[0].contains("raw literal")
        );
    }

    #[test]
    fn rejects_malformed_and_duplicate_typed_domains() {
        let directory = repository("duplicate-domain");
        directory.write(
            "crates/owner/src/lib.rs",
            "const BAD_DOMAIN: FingerprintDomain = FingerprintDomain::new(\"nia.owner..product.v2\");\nconst FIRST_DOMAIN: FingerprintDomain = FingerprintDomain::new(\"nia.owner.product.v2\");\nconst SECOND_DOMAIN: FingerprintDomain = FingerprintDomain::new(\"nia.owner.product.v2\");\n",
        );
        assert!(fingerprint_domain_errors(directory.path()).unwrap()[0].contains("owner..product"));
        assert!(
            typed_fingerprint_domain_errors(directory.path())
                .unwrap()
                .iter()
                .any(|error| error.contains("duplicate"))
        );
    }

    #[test]
    fn rejects_registered_magic_outside_registry() {
        let directory = repository("duplicate-magic");
        directory.write(
            "crates/owner/src/lib.rs",
            "const MAGIC: &[u8; 8] = b\"NIAFMT01\";\n",
        );
        assert!(
            global_identity_errors(directory.path()).unwrap()[0].contains("outside nia-compat")
        );
    }

    #[test]
    fn rejects_workspace_version_outside_authorities() {
        let directory = repository("duplicate-version");
        directory.write("README.md", "current version: 1.2.3\n");
        assert!(release_version_errors(directory.path()).unwrap()[0].contains("README.md"));
    }
}
